File: web/src/awards/wcsa.ts
import schoolList from '../data/wcsa.json'
import type { Qso, WcsaLevel, WcsaResult, WcsaSlot } from '../types'

const SCHOOLS = Object.fromEntries(
  Object.entries(schoolList).map(([call, name]) => [call.toUpperCase(), String(name)]),
) as Record<string, string>

const AWARD_LEVELS: Array<{ name: WcsaLevel['name']; requiredSchools: number; requiredSlots: number }> = [
  { name: 'Bronze', requiredSchools: 1, requiredSlots: 3 },
  { name: 'Silver', requiredSchools: 2, requiredSlots: 6 },
  { name: 'Gold', requiredSchools: 3, requiredSlots: 10 },
]

function isWcsaQslReceived(qso: Qso): boolean {
  if (qso.QSL_RCVD?.toUpperCase() === 'Y') return true
  if (qso.LOTW_QSL_RCVD?.toUpperCase() === 'Y') return true
  return false
}

function normalizeQso(qso: Qso): Qso {
  const normalized: Record<string, string> = {}
  for (const [key, value] of Object.entries(qso)) {
    normalized[key.toUpperCase()] = value === undefined ? '' : String(value).trim()
  }

  for (const key of ['CALL', 'MODE', 'PROP_MODE', 'SAT_NAME']) {
    if (normalized[key]) normalized[key] = normalized[key].toUpperCase()
  }

  return normalized as Qso
}

function getSchoolCall(call: string): string | null {
  const normalizedCall = call.trim().toUpperCase()
  if (normalizedCall in SCHOOLS) return normalizedCall

  for (const part of normalizedCall.split('/')) {
    if (part in SCHOOLS) return part
  }

  return null
}

function getSlotKey(qso: Qso): string | null {
  const call = qso.CALL?.toUpperCase() || ''
  const mode = qso.MODE?.toUpperCase() || ''
  const propMode = qso.PROP_MODE?.toUpperCase() || ''

  if (!call || !qso.QSO_DATE || !mode) return null

  if (propMode === 'SAT') {
    const satName = qso.SAT_NAME?.toUpperCase() || ''
    if (!satName) return null
    return [call, 'SAT', satName, mode].join('|')
  }

  const band = qso.BAND?.toUpperCase() || ''
  if (!band) return null

  if (propMode === 'EME' || propMode === 'MS') {
    return [call, band, mode, propMode].join('|')
  }

  return [call, band, mode].join('|')
}

function qsoStamp(qso: Pick<Qso, 'QSO_DATE' | 'TIME_ON'>): string {
  return `${qso.QSO_DATE || ''}${qso.TIME_ON || ''}`
}

function toWcsaSlot(schoolCall: string, qso: Qso): WcsaSlot {
  return {
    schoolCall,
    schoolName: SCHOOLS[schoolCall],
    qso: {
      CALL: qso.CALL,
      QSO_DATE: qso.QSO_DATE,
      TIME_ON: qso.TIME_ON || '',
      BAND: qso.BAND || '',
      MODE: qso.MODE,
      PROP_MODE: qso.PROP_MODE || '',
      SAT_NAME: qso.SAT_NAME || '',
    },
  }
}

function buildResult(slots: Map<string, WcsaSlot>): WcsaResult {
  const schoolCalls = Array.from(new Set(Array.from(slots.values()).map((slot) => slot.schoolCall))).sort()
  const slotCountsBySchool: Record<string, number> = {}
  for (const schoolCall of schoolCalls) slotCountsBySchool[schoolCall] = 0
  for (const slot of slots.values()) slotCountsBySchool[slot.schoolCall]++

  const schoolCount = schoolCalls.length
  const slotCount = slots.size
  const levels = AWARD_LEVELS.map((level) => ({
    ...level,
    achieved: schoolCount >= level.requiredSchools && slotCount >= level.requiredSlots,
  }))

  return {
    slots,
    schoolCount,
    slotCount,
    levels,
    slotCountsBySchool,
  }
}

export function computeWcsa(qsos: Qso[]): WcsaResult {
  const slots = new Map<string, WcsaSlot>()

  for (const qso of qsos) {
    const normalizedQso = normalizeQso(qso)
    const schoolCall = getSchoolCall(normalizedQso.CALL || '')

    if (!schoolCall) continue
    if (!isWcsaQslReceived(normalizedQso)) continue

    const slotKey = getSlotKey(normalizedQso)
    if (!slotKey) continue

    const existing = slots.get(slotKey)
    if (!existing || qsoStamp(normalizedQso) < qsoStamp(existing.qso)) {
      slots.set(slotKey, toWcsaSlot(schoolCall, normalizedQso))
    }
  }

  return buildResult(slots)
}

File: python/awards/wcsa.py
import json
import sys

from common.qsl import is_qsl_received


AWARD_LEVELS = [
    ('Bronze', 1, 3),
    ('Silver', 2, 6),
    ('Gold', 3, 10),
]


def load_school_list(path_to_school_list):
    with open(path_to_school_list, 'r', encoding='utf-8') as f:
        schools = json.load(f)

    return schools


def compute_wcsa(qsos, schools, warning_stream=sys.stderr):
    normalized_schools = {call.upper(): name for call, name in schools.items()}
    slots = {}

    for qso in qsos:
        normalized_qso = _normalize_qso(qso)
        school_call = _get_school_call(normalized_qso.get('CALL', ''), normalized_schools)

        if not school_call:
            continue
        if not is_qsl_received(normalized_qso):
            continue

        slot_key = _get_slot_key(normalized_qso, warning_stream)
        if slot_key is None:
            continue

        existing = slots.get(slot_key)
        if existing is None or _qso_stamp(normalized_qso) < _qso_stamp(existing['QSO']):
            slots[slot_key] = {
                'SCHOOL_CALL': school_call,
                'SCHOOL_NAME': normalized_schools[school_call],
                'QSO': normalized_qso,
            }

    return _build_result(slots)


def _normalize_qso(qso):
    normalized_qso = {}
    for key, value in qso.items():
        normalized_qso[str(key).upper()] = '' if value is None else str(value).strip()

    for key in ('CALL', 'MODE', 'PROP_MODE', 'SAT_NAME'):
        if key in normalized_qso:
            normalized_qso[key] = normalized_qso[key].upper()

    return normalized_qso


def _get_school_call(call, schools):
    normalized_call = call.strip().upper()
    if normalized_call in schools:
        return normalized_call

    for part in normalized_call.split('/'):
        if part in schools:
            return part

    return None


def _get_slot_key(qso, warning_stream):
    call = qso.get('CALL', '').upper()
    prop_mode = qso.get('PROP_MODE', '').upper()

    required_fields = ['CALL', 'QSO_DATE', 'MODE']
    if prop_mode == 'SAT':
        required_fields.append('SAT_NAME')
    else:
        required_fields.append('BAND')

    missing_fields = [field for field in required_fields if not qso.get(field)]
    if missing_fields:
        print(
            "WARNING: QSO with %s is missing required field(s): %s, skipped" % (
                call or 'UNKNOWN',
                ', '.join(missing_fields),
            ),
            file=warning_stream,
        )
        return None

    mode = qso['MODE'].upper()
    if prop_mode == 'SAT':
        return (call, 'SAT', qso['SAT_NAME'].upper(), mode)
    if prop_mode in ('EME', 'MS'):
        return (call, qso['BAND'].upper(), mode, prop_mode)
    return (call, qso['BAND'].upper(), mode)


def _build_result(slots):
    school_calls = sorted({slot['SCHOOL_CALL'] for slot in slots.values()})
    slot_counts_by_school = {}
    for school_call in school_calls:
        slot_counts_by_school[school_call] = 0
    for slot in slots.values():
        slot_counts_by_school[slot['SCHOOL_CALL']] += 1

    school_count = len(school_calls)
    slot_count = len(slots)
    levels = []
    for name, required_schools, required_slots in AWARD_LEVELS:
        levels.append({
            'NAME': name,
            'REQUIRED_SCHOOLS': required_schools,
            'REQUIRED_SLOTS': required_slots,
            'ACHIEVED': school_count >= required_schools and slot_count >= required_slots,
        })

    return {
        'SLOTS': slots,
        'SCHOOL_COUNT': school_count,
        'SLOT_COUNT': slot_count,
        'LEVELS': levels,
        'SLOT_COUNTS_BY_SCHOOL': slot_counts_by_school,
    }


def _qso_stamp(qso):
    return qso.get('QSO_DATE', '') + qso.get('TIME_ON', '')
